client/state/audio/middleware.js

```diff
@@ -0,0 +1,64 @@
+
+/**
+ * Internal dependencies
+ */
+import {
+	HAPPYCHAT_RECEIVE_EVENT,
+	AUDIO_PLAY,
+} from 'state/action-types';
+import { audioPlay } from './actions';
+
+// Maps the audio reference to the audio file source.
+// In future we might want to store this in Redux if
+// we want the sounds to be configurable.
+const REFERENCE_MAP = {
+	'happychat-message-received': '/calypso/audio/chat-pling.wav',
+};
+
+/**
+ * Plays a sound when a message is received on Happychat
+ * @param {function} dispatch Redux dispatcher
+ * @param {Object} action Redux action
+ */
+export const dispatchHappychatSound = ( dispatch, { event } ) => {
+	// Don't play a sound when the customer sends a message
+	if ( event.source === 'customer' ) {
+		return;
+	}
+
+	dispatch( audioPlay( 'happychat-message-received' ) );
+};
+
+export const playSound = ( dispatch, { reference } ) => {
+	if ( typeof Audio !== 'function' ) {
+		// No Audio support in this browser
+		return;
+	}
+
+	const audio = new Audio( REFERENCE_MAP[ reference ] );
+	audio.play();
+};
+
+/**
+ * Action Handlers
+ */
+
+export const handlers = {
+	// Actions which trigger a sound
+	[ HAPPYCHAT_RECEIVE_EVENT ]: dispatchHappychatSound,
+
+	// Actually plays the sound
+	[ AUDIO_PLAY ]: playSound,
+};
+
+/**
+ * Middleware
+ */
+
+export default ( { dispatch, getState } ) => ( next ) => ( action ) => {
+	if ( handlers.hasOwnProperty( action.type ) ) {
+		handlers[ action.type ]( dispatch, action, getState );
+	}
```

Comment:
>I'm curious where the theory comes from?

The optimization comes from the lookup operator. `hasOwnProperty` can be slow (I imagine `has` is even slower):

https://measurethat.net/Benchmarks/Show/1214/0/undefined-vs-typeof-vs-in-vs-hasownproperty

Specifically in the previous discussion Firefox was showing to be particularly slow on `hasOwnProperty`:

https://github.com/Automattic/wp-calypso/pull/8606#discussion_r92007182

The need for `Object.create` vs. `{}` is to avoid operators other than `hasOwnProperty` checking an object's prototype. `Object.create` creates an object without any prototype members (e.g. compare `'valueOf' in {}` vs. `'valueOf in Object.create( null )`).

This is both an edge case and a micro-optimization, but I'm particularly sensitive to middlewares being as fast as possible (because we dispatch so much). In fact, I'd much rather we have as _few_ middlewares as possible. And especially because this pattern (looking up handlers in an object) is becoming so common, I wonder if we should think about consolidating them.